### PR TITLE
[Clang][AArch64] Require SVE or SSVE for scalable types.

### PR DESCRIPTION
**Pull request by sdesmalen-arm**
Scalable types are only available when:
* The function is compiled with +sve
* The function is compiled with +sme and the function is executed in Streaming-SVE mode.

### CODE REVIEW

**Review by aemerson — APPROVED** [review]

**Review by paulwalker-arm — CHANGES_REQUESTED** [review]
As discussed offline, I don't think we want to be this strict.  As demonstrated by the changes to the ACLE tests, this change makes it impossible to distribute a library in binary form that can work for both SVE and InStreamingMode environments.  I believe functions decorated with `__arm_streaming_compatible` should be allowed to assume the presence of the subset of instructions that is available to both environments. Library users get protected at the point they call such functions whereby a compilation error is emitted when the caller either doesn't have access to SVE or is not in streaming mode.

File: clang/lib/Sema/SemaDecl.cpp [hunk 1]
```diff
@@ -8982,11 +8982,18 @@ void Sema::CheckVariableDeclarationType(VarDecl *NewVD) {
     const FunctionDecl *FD = cast<FunctionDecl>(CurContext);
     llvm::StringMap<bool> CallerFeatureMap;
     Context.getFunctionFeatureMap(CallerFeatureMap, FD);
-    if (!Builtin::evaluateRequiredTargetFeatures(
-        "sve", CallerFeatureMap)) {
-      Diag(NewVD->getLocation(), diag::err_sve_vector_in_non_sve_target) << T;
-      NewVD->setInvalidDecl();
-      return;
```
Comment (paulwalker-arm) [inline clang/lib/Sema/SemaDecl.cpp:8989]: Is dropping the immediate return upon setting up a diagnostic intentional?

File: clang/test/CodeGen/aarch64-sme2-intrinsics/acle_sme2_reinterpret_svcount_svbool.c [hunk 2]
```diff
@@ -9,6 +9,12 @@
 
 #include <arm_sme.h>
 
+#if defined __ARM_FEATURE_SME
+#define MODE_ATTR __arm_streaming
+#else
+#define MODE_ATTR __arm_streaming_compatible
```
Comment (paulwalker-arm) [inline clang/test/CodeGen/aarch64-sme2-intrinsics/acle_sme2_reinterpret_svcount_svbool.c:15]: Do you need to use `__arm_streaming_compatible` here?  Now we've agreed this keyword has no effect on the target features in use I think `MODE_ATTR` should remain blank to mirror the expected usage when SME is not in use.

**Review by paulwalker-arm — COMMENTED** [review]
After further discussion I now understand the `__arm_streaming_compatible` keyword has no affect on the target features in play and only tells the compiler not to emit any SM state changing instructions as part of the calling convention.

File: clang/lib/Sema/SemaDecl.cpp (hunk 1 above)
Comment (sdesmalen-arm) [reply clang/lib/Sema/SemaDecl.cpp:8989]: It was not, thanks.

**Review by paulwalker-arm — APPROVED** [review]

**Pull request merged**